x/ibc/03-connection/types/version.go

```diff
@@ -1,96 +1,171 @@
+)
+
+// GetCompatibleVersions returns a descending ordered set of compatible IBC
+// versions for the caller chain's connection end. The latest supported
+// version should be first element and the set should descend to the oldest
+// supported version.
 func GetCompatibleVersions() []string {
-	return []string{"1.0.0"}
+	return []string{DefaultIBCVersion}
 }
 
-// LatestVersion gets the latest version of a connection protocol
+// CreateVersionString constructs a valid connection version given a
+// version identifier and feature set. The format is written as:
 //
-// CONTRACT: version array MUST be already sorted.
-func LatestVersion(versions []string) string {
-	if len(versions) == 0 {
-		return ""
-	}
-	return versions[len(versions)-1]
+// ([version_identifier],[feature_0,feature_1,feature_2...])
+//
+// A connection version is considered invalid if it is not in this format
+// or violates one of these rules:
+// - the version identifier is empty or contains commas
+// - a specified feature contains commas
+func CreateVersionString(identifier string, featureSet []string) string {
+	return fmt.Sprintf("(%s,%s)", identifier, featureSet)
 }
 
-// PickVersion picks the counterparty latest version that is matches the list
-// of compatible versions for the connection.
-func PickVersion(counterpartyVersions, compatibleVersions []string) string {
+// UnpackVersion parses a version string and returns the identifier and the
+// feature set of a version. An error is returned if the version is not valid.
+func UnpackVersion(version string) (string, []string, error) {
+	// validate version so valid splitting assumptions can be made
+	if err := host.ConnectionVersionValidator(version); err != nil {
+		return "", nil, err
+	}
 
-	n := len(counterpartyVersions)
-	m := len(compatibleVersions)
+	// peel off prefix and suffix of the tuple
+	version = strings.TrimPrefix(version, "(")
+	version = strings.TrimSuffix(version, ")")
 
-	// aux hash maps to lookup already seen versions
-	counterpartyVerLookup := make(map[string]bool)
-	compatibleVerLookup := make(map[string]bool)
+	// split into identifier and feature set
+	splitVersion := strings.SplitN(version, ",", 2)
+	identifier := splitVersion[0]
 
-	// versions suported
-	var supportedVersions []string
+	// peel off prefix and suffix of features
+	featureSet := strings.TrimPrefix(splitVersion[1], "[")
+	featureSet = strings.TrimSuffix(featureSet, "]")
 
-	switch {
-	case n == 0 || m == 0:
-		return ""
-	case n == m:
-		for i := n - 1; i >= 0; i-- {
-			counterpartyVerLookup[counterpartyVersions[i]] = true
-			compatibleVerLookup[compatibleVersions[i]] = true
+	// check if features are empty
+	if len(featureSet) == 0 {
+		return identifier, []string{}, nil
+	}
 
-			// check if we've seen any of the versions
-			if _, ok := compatibleVerLookup[counterpartyVersions[i]]; ok {
-				supportedVersions = append(supportedVersions, counterpartyVersions[i])
-			}
+	features := strings.Split(featureSet, ",")
```

Comment:
missing check `features[0] == featureSet`. See the `Split` func godoc for reference

Reply:
> If s does not contain sep and sep is not empty, Split returns a slice of length 1 whose only element is s.

Reply:
I think this is fine? I have the check if its empty, and then either it is size one and that feature is returned as []string{feature}, or it contains sep and it is returned as []string{feature0,feature1, etc}. I'm not sure what that check would result in if true